### PR TITLE
Call community-api with the tier update

### PR DESCRIPTION
**Pull request by markreesmoj**

### CODE REVIEW

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppstier/controller/TierCalculationController.kt [hunk 1]
```diff
@@ -26,6 +26,6 @@ class TierCalculationController(private val tierCalculationService: TierCalculat
   )
   @GetMapping("crn/{crn}/tier")
   fun getNotifications(@PathVariable(required = true) crn: String): ResponseEntity<TierDto> {
-    return ResponseEntity.ok(tierCalculationService.getTierByCrn(crn))
+    return ResponseEntity.ok(tierCalculationService.getOrCalculateTierByCrn(crn))
```
Comment (markreesmoj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppstier/controller/TierCalculationController.kt:29]: This is just renaming the method - it does the same as it always did

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppstier/dto/CalculationResultDto.kt [hunk 2]
```diff
@@ -0,0 +1,5 @@
+package uk.gov.justice.digital.hmpps.hmppstier.dto
+
+data class CalculationResultDto constructor(
+  val tierDto: TierDto?
+)
```
Comment (markreesmoj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppstier/dto/CalculationResultDto.kt:5]: This is not very nice. The requirement is to check whether the calculation has changed, and the difficulty is when there is no previous calculation. I haven't found an elegant way of doing this and I didn't want to change the TierDto

File: src/test/kotlin/uk/gov/justice/digital/hmpps/hmppstier/integration/MockedEndpointsTestBase.kt [hunk 3]
```diff
@@ -31,6 +34,13 @@ abstract class MockedEndpointsTestBase : IntegrationTestBase() {
     mockAssessmentApiServer.stop()
   }
 
+  fun calculationMessage(crn: String): String {
+    val validMessage: String =
+      Files.readString(Paths.get("src/test/resources/fixtures/sqs/tier-calculation-event.json")).replace("X373878", crn)
+
+    return validMessage
+  }
+
```
Comment (markreesmoj) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/hmppstier/integration/MockedEndpointsTestBase.kt:43]: making it easier to write outside-in tests where we just call the listener

File: src/test/kotlin/uk/gov/justice/digital/hmpps/hmppstier/integration/TierCalculationTest.kt [hunk 4]
```diff
@@ -1,204 +1,218 @@
...
-import uk.gov.justice.digital.hmpps.hmppstier.service.TierCalculationService
+import uk.gov.justice.digital.hmpps.hmppstier.service.TierCalculationRequiredEventListener
 
-@TestInstance(TestInstance.Lifecycle.PER_CLASS)
+@TestInstance(PER_CLASS)
 class TierCalculationTest : MockedEndpointsTestBase() {
 
   @Autowired
-  lateinit var service: TierCalculationService
+  lateinit var listener: TierCalculationRequiredEventListener
```
Comment (markreesmoj) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/hmppstier/integration/TierCalculationTest.kt:16]: Now more of an outside-in test where we call the listener and check that the update happens as expected. This makes it a bit easier to refactor the service - changing the return types for example made a real mess of this

If/when we get requirements for the controller, this kind of test would be easier to write against that.

**Review by markreesmoj — COMMENTED** [review]
reviewed by accident

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppstier/dto/CalculationResultDto.kt (hunk 2 above)
Comment (nickmcmahon01) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppstier/dto/CalculationResultDto.kt:5]: Is this not a case of pulling the existing tier out of the database, calculating the new one, and comaparing
then 
`
old = getOld(crn)
new = calculateNew(crn)
if (old != new) { writebacktodelius() }`

Comment (markreesmoj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppstier/dto/CalculationResultDto.kt:5]: yes, as long as the old one is there. I didn't want to expose the DB entity in the listener though. 

Ultimately there's quite a lot going on in the listener, probably need to push some of it down

Comment (markreesmoj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppstier/dto/CalculationResultDto.kt:5]: The other complication is supporting the controller, for which we currently have no specs. What the listener needs in order to call the updater does not have a lot in common with what the controller returns

File: src/test/kotlin/uk/gov/justice/digital/hmpps/hmppstier/service/TierCalculationServiceTest.kt [hunk 5]
```diff
@@ -162,7 +165,7 @@ internal class TierCalculationServiceTest {
       val tier = service.calculateTierForCrn(crn)
       standardVerify()
 
-      assertThat(tier.data.protect.points).isEqualTo(20)
```
Comment (markreesmoj) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/hmppstier/service/TierCalculationServiceTest.kt:165]: Every time I change anything, I seem to have to spend 20 minutes in this test despite none of the functionality changing

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppstier/client/CommunityApiClient.kt [hunk 6]
```diff
@@ -69,6 +71,13 @@ class CommunityApiClient(@Qualifier("communityWebClientAppScope") private val we
       .bodyToMono(Requirements::class.java)
       .block()?.requirements ?: listOf()
   }
+
+  fun updateTier(tierDto: TierDto, crn: String): ResponseEntity<Void>? {
+    return webClient
+      .post()
+      .uri("offenders/crn/$crn/tier/${tierDto.protectLevel}${tierDto.changeLevel.ordinal}")
+      .retrieve().toBodilessEntity().block()
```
Comment (markreesmoj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppstier/client/CommunityApiClient.kt:79]: This endpoint does return the offender but I suppose we aren't interested in that

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppstier/service/TierCalculationService.kt [hunk 7]
```diff
@@ -17,15 +18,27 @@ class TierCalculationService(
...
       log.info("Returned tier for $crn")
     }
   }
 
-  fun calculateTierForCrn(crn: String): TierCalculationEntity {
+  fun calculateTierForCrn(crn: String): CalculationResultDto {
+    val existingCalculation = getLatestTierCalculation(crn)
+    val newTier = calculateTier(crn)
+    val isUpdated: Boolean
+    if (existingCalculation == null) {
```
Comment (markreesmoj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppstier/service/TierCalculationService.kt:32]: Kotlin magic needed here

**Review by nickmcmahon01 — APPROVED** [review]

**Pull request merged**